daisi/src/minhton-ns3/scenario_steps.h

```diff
@@ -0,0 +1,207 @@
+
+#ifndef DAISI_MINHTON_NS3_MINHTON_SCENARIO_STEPS_H_
+#define DAISI_MINHTON_NS3_MINHTON_SCENARIO_STEPS_H_
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+#include "manager/scenariofile_component.h"
+
+#define TYPE_NAME(name) \
+  static std::string typeName() { return name; }
+
+namespace daisi::minhton_ns3 {
+
+struct JoinOne {
+  TYPE_NAME("join-one")
+
+  std::optional<uint32_t> delay;
+  std::optional<uint32_t> level;
+  std::optional<uint32_t> number;
+  std::optional<uint32_t> index;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(index);
+  }
+  static std::string getTypeName() { return "join-one"; }
+};
+
+struct JoinMany {
+  TYPE_NAME("join-many")
+
+  std::optional<uint32_t> delay;
+  uint32_t number;
+  std::string mode;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(mode);
+  }
+};
+
+struct LeaveOne {
+  TYPE_NAME("leave-one")
+
+  std::optional<uint32_t> delay;
+  std::optional<uint32_t> level;
+  std::optional<uint32_t> number;
+  std::optional<uint32_t> index;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(index);
+  }
+};
+
+struct LeaveMany {
+  TYPE_NAME("leave-many");
+  std::optional<uint32_t> delay;
+  uint32_t number;
+  std::string mode;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(mode);
+  }
+};
+
+struct SearchMany {
+  TYPE_NAME("search-many");
+
+  std::optional<uint32_t> delay;
+  uint32_t number;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(number);
+  }
+};
+
+struct SearchAll {
+  TYPE_NAME("search-all");
+
+  std::optional<uint32_t> delay;
+
+  void parse(YAML::Node node) { SERIALIZE_VAR(delay); }
+};
+
+struct FailOne {
+  TYPE_NAME("fail-one");
+
+  std::optional<uint32_t> delay;
+  uint32_t number;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(number);
+  }
+};
+
+struct FailMany {
+  TYPE_NAME("fail-many");
+
+  std::optional<uint32_t> delay;
+  uint32_t level;
+  uint32_t number;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+  }
+};
+
+struct MixedExecution {
+  TYPE_NAME("mixed-execution");
+
+  std::optional<uint32_t> delay;
+  uint32_t join_number;
+  uint32_t leave_number;
+  uint32_t search_number;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(join_number);
+    SERIALIZE_VAR(leave_number);
+    SERIALIZE_VAR(search_number);
+  }
+};
+
+struct ValidateLeave {
+  TYPE_NAME("validate-leave");
+
+  std::optional<uint32_t> delay;
+
+  void parse(YAML::Node node) { SERIALIZE_VAR(delay); }
+};
+
+struct FindQuery {
+  TYPE_NAME("find-query");
+
+  uint64_t level;
+  uint64_t number;
```

Comment:
Should probably be `uint32_t` just like in the other structs. And maybe think about whether validity_threshold actually needs to be a `uint64_t`.

Reply:
Think big :wink: 
No, I just changed this to ``uint32_t``. ``uint64_t`` was used by the old parser for all integer types, but 32-bit wide integers should be fine.